### ISSUE TITLE
Modify Tag Manager module setup with Analytics guidance

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#1381]
## Feature Description

If the Tag Manager container a user selects includes an Analytics tag, the user should be guided to deal with that properly.

**PR for this should be opened against `feature/ga-gtm-separation`.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

When the user selects a GTM container, Site Kit should check whether there's a GA tag configured for that container, via the `getLiveContainerVersion` selector.

Based on this detection, we need the following logic to happen:

### No GA tag in GTM, GA module not active

* Don't do anything (other than what's happening already).

### No GA tag in GTM, GA module is active

* Don't do anything (other than what's happening already).

### GA tag in GTM, GA module not active
**The changes in this section specifically should only apply to GTM set up, not GTM settings.**
* Recommend the user to also activate the GA module:  
	> Looks like you're already using Google Analytics within your Google Tag Manager configuration.
	> Activate the Google Analytics module in Site Kit to see relevant insights in your dashboard.
* Change the primary button for current functionality into a secondary button that says **"Complete setup without Analytics"**. Functionality should remain as-is.
* Add new primary submit button to say **"Continue to Analytics setup"**.  
When clicked, complete setup and activate GA module so that user gets redirected to OAuth consent screen for GA scopes and then GA module setup.

#### Property ID in GA tag in GTM and GA module match

* Inform the user:  
	> "Looks like you're using Google Analytics. Your Analytics property UA-1234567-1 is already set up in your Google Tag Manager configuration, so Site Kit will switch to using Google Tag Manager for Analytics."
* When user clicks button to complete setup, the GA module's `useSnippet` option should be set to `false`.

#### Property ID in GA tag in GTM and GA module match (GA snippet placed by other plugin/theme)

* Block the user from proceeding (since we can't prevent the existing tag from being output):  
	> "Looks like you're using Google Analytics within your Google Tag Manager configuration. [...]"

#### Property ID in GA tag in GTM and GA module don't match

* Block the user from proceeding:
	> "Looks like you're already using Google Analytics within your Google Tag Manager configuration. However, its Analytics property UA-1234568-1 is different from the Analytics property UA-1234567-1, which was previously selected in the plugin. You need to configure the same Analytics property in both places."

### How to deal with primary AMP

* Logic should be exactly the same as for non-AMP - just against the `ampContainerID` instead of the `containerID`.

### How to deal with secondary AMP

* If both the non-AMP and the AMP container have no GA property or if they have the same GA property, the same logic as above should apply.
* Otherwise, block the user from proceeding:
	> "Looks like you're using Google Analytics within your Google Tag Manager configuration. However, your AMP container uses a different configuration from your non-AMP container. Make sure you set up the same Google Analytics property for both of them." (this is preliminary messaging, may be refined, but let's go with this for now)

## Implementation Brief

Note the use of `find` here should not be used verbatim as it's intended for showing how to look up the necessary object in a list.
`undefined` should only be returned for the case where loading/resolution is still in progress. An object that is non-existent or not present should be returned as `null`.

Also note that the use of the word "tag" here is specific to GTM containers and does not refer to a container ID or property ID from a code snippet as it is sometimes referred to elsewhere.

* Add a new `getLiveContainerVariable( accountID, internalContainerID, variableName )` registry selector  
	Gets variable object by name for the given live container
	- `variableName` should have any leading or trailing braces removed from the input `variableName` before searching
	- selects `getLiveContainer( accountID, internalContainerID )`
	- returns `{(Object|null|undefined)}` `liveContainer.variable.find( ( variable ) => variable.name === variableName ) || null` if `response.variable` is present (it may not be)
* Add a new `getLiveContainerAnalyticsTag( accountID, internalContainerID )` registry selector  
	Gets container tag object for Analytics from the given live container, if present
	- `getLiveContainer( accountID, internalContainerID )`
	- returns `{(Object|null|undefined)}` `response.tag.find( ( tag ) => tag.type === 'ua' ) || null` if the live container is loaded, otherwise `undefined`.
* Add a new `hasliveContainerAnalyticsTag( accountID, internalContainerID )` registry selector
	- `getLiveContainerAnalyticsTag( accountID, internalContainerID )`
	- Returns `true` response is a plain object, otherwise `false`; `undefined` if not loaded yet
* Add a new `getLiveContainerAnalyticsPropertyID( accountID, internalContainerID )` registry selector
	- `getLiveContainerAnalyticsTag( accountID, internalContainerID )`  
	returns the `value` of the `gaTag.parameter` with the key of `gaSettings`  
	_Note that the `value` can be a variable, in which case it needs to be looked up by name (without the surrounding curly braces; `{{Variable Name}}` is `Variable Name`)_
	- If `value` is a variable, call `getLiveContainerVariable( accountID, internalContainerID, value )`
		- return the `value` of the `gaVariable.parameter` with the key of `trackingId`, `null` if not found, or `undefined` if not loaded yet.
* Add a new `getAnalyticsPropertyIDs` as a registry selector  
	Returns `Array` of found property IDs (without duplicates) from GTM Analytics tags, or `undefined` if not loaded yet
* Add a new `hasMultipleAnalyticsPropertyIDs`  as a registry selector  
	Returns `true` if `getAnalyticsPropertyIDs` has multiple _or_ if `getAnalyticsPropertyIDs` has one while one of the GTM containerIDs are currently set to `container_create` or `undefined` if not loaded yet
* Add a new `getSingleAnalyticsPropertyID` as a registry selector  
	Returns the first item (string) from `getAnalyticsPropertyIDs`, but only if `! hasMultipleAnalyticsPropertyIDs` (bool) `false`; `undefined` if not loaded yet
* ~~Add a new `hasSingleAnalyticsPropertyID` as a registry selector  ~~
	Returns `!! getSingleAnalyticsPropertyID` or `undefined` if not loaded yet
* Add a new `hasAnyAnalyticsPropertyID` as a registry selector  
	Returns `true` if `getAnalyticsPropertyIDs` contains any non-`null` ID, or `undefined` if not loaded yet
* Update `selectors.canSubmitChanges` to return `false` if any of the following are true (user must choose a different container or abort setup):
	- `hasMultipleAnalyticsPropertyIDs()` is `undefined` or `true`
	- `getSingleAnalyticsPropertyID()` is a valid property ID and does not match an the current property ID for Analytics
* Add `googlesitekit-modules-analytics` asset as a dependency of `googlesitekit-modules-tagmanager`
* Add `googlesitekit/v1/modules/analytics/data/settings` to preloaded requests (saves extra request if GA is not active - no change if it is)

**Variable references for below**
* `select( core/modules ).isModuleActive( 'analytics' )` as `analyticsModuleActive`
* `select( modules/analytics ).getAdminReauthURL()` as `analyticsModuleReauthURL`
* `select( modules/analytics ).getPropertyID()` as `analyticsPropertyID`
* `select( modules/tagmanager ).getSingleAnalyticsPropertyID()` as `gtmAnalyticsPropertyID`
* `select( modules/tagmanager ).hasMultipleAnalyticsPropertyIDs()` as `hasMultipleAnalyticsPropertyIDs`

#### `FormInstructions`

* If `gtmAnalyticsPropertyID` and `gtmAnalyticsPropertyID === analyticsPropertyID`
	- Display message from AC "Your Analytics property UA-1234567-1 is already set up in your Google Tag Manager configuration..."
* If `hasMultipleAnalyticsPropertyIDs`  
	- Display error message from AC (user will be blocked by logic in `canSubmitChanges`)
	> Looks like you're already using Google Analytics within your Google Tag Manager configuration [...]

#### `SetupForm`
* `setIsNavigating` should be additionally passed as a prop from `SetupMain`
* After `submitChanges` is called and `! error`, if `gtmAnalyticsPropertyID` and `analyticsModuleActive`
	- `dispatch( modules/analytics ).setUseSnippet( false )`
	- `dispatch( modules/analytics ).saveSettings()` on submit, before calling `finishSetup`
* If `gtmAnalyticsPropertyID` and `! analyticsModuleActive`
- Render alternate submit buttons:
	- (Primary) **Continue to Analytics setup**  
		- Submits changes directly (without calling `finishSetup`)
		- activates Analytics via `dispatch( core/modules ).activateModule( 'analytics' )`
		- call `setIsNavigating( true )`
		- navigates to `analyticsModuleReauthURL`
	- (Secondary) **Complete setup without Analytics**
	Submits form as usual
* After `submitChanges` is called and `! error`, if `gtmAnalyticsPropertyID` and `analyticsModuleActive`
	- `dispatch( modules/analytics ).setUseSnippet( false )`
	- `dispatch( modules/analytics ).saveSettings()`

#### `SettingsEdit`

* After `submitChanges` is called and `! error`, if `gtmAnalyticsPropertyID` and `analyticsModuleActive`
	- `dispatch( modules/analytics ).setUseSnippet( false )`
	- `dispatch( modules/analytics ).saveSettings()`

## QA Brief

> **Ensure testing is done using the latest `feature/ga-gtm-separation` branch**

**Prerequisites**
* Analytics
	- a few different properties (can be same account)
* Tag Manager
	- a web container with no GA tag
	- an AMP container with no GA tag
	- a web container with a GA tag matching a property ID in your Analytics account
	- an AMP container with a GA tag matching a property ID in your Analytics account
	- ensure all containers used (and any updates) are published
* AMP plugin installed

### QA matrix

All cases are in the context of the Tag Manager module setup screen only. The settings view is unchanged by this.

| Analytics Module | AMP Mode | Web Container | AMP Container | Expected |
| --- | --- | --- | --- | --- |
| Inactive | Inactive | No GA | N/A | ✅  No change (same as before) |
| Inactive | Inactive | With GA property | N/A | ✅  Primary action is continue with Analytics setup, secondary continue without |
| Setup | Inactive | With same GA property as Analytics | N/A | ✅  See message including "Site Kit will switch to using Google Tag Manager for Analytics", single button to continue |
| Setup | Inactive | With different GA property than Analytics | N/A | ❌  See message including "You need to configure the same Analytics property in both places", blocked from continuing |
| Inactive | Primary | N/A | No GA | ✅  No change (same as before) |
| Inactive | Primary | N/A | With GA property | ✅  Primary action is continue with Analytics setup, secondary continue without |
| Setup | Primary | N/A | With same GA property as Analytics | ✅  See message including "Site Kit will switch to using Google Tag Manager for Analytics", single button to continue |
| Setup | Primary | N/A | With different GA property than Analytics | ❌  See message including "You need to configure the same Analytics property in both places", blocked from continuing |
| Inactive | Secondary | No GA | No GA | ✅ No change (same as before) |
| Inactive | Secondary | With same GA property | With same GA property | ✅  Primary action is continue with Analytics setup, secondary continue without |
| Setup | Secondary | With same GA property as Analytics | With same GA property as Analytics | ✅  See message including "Site Kit will switch to using Google Tag Manager for Analytics", single button to continue |
| Setup | Secondary | With different GA property than Analytics | With different GA property than Analytics | ❌  See message including "You need to configure the same Analytics property in both places", blocked from continuing |
| Setup | Secondary | With GA property A | With GA property B | ❌  See message including "You need to configure the same Analytics property in both containers", blocked from continuing |

## Changelog entry

* Provide guidance in Tag Manager setup flow when container is configured to use Analytics, including checks to ensure an eventual AMP container behaves correctly as well.

**Comment by marrrmarrr** [comment]
Made some edits to the language. LGTM with these edits.

**Comment by aaemnnosttv** [comment]
@felixarntz I have a WIP IB going here, but there are a few issues that need working out.

1. Things are more complicated for `SettingsEdit` as we don't have control over the main submit buttons (apart from the disabled state).
1. The complexity is particularly compounded for cases using secondary AMP where they have 2 different containers selected. I'm not sure if those cases have been considered here, but how should we handle those? What if one container has a GA tag and the other does not? What if they both have GA tags with non-matching property IDs (is that normal for AMP/web contexts to use different properties?)
1. It seems like we need to load the Analytics datastore as a dependency for this to work as well as it needs to be able to read/write settings, check for an existing tag, and get the reAuthURL. I think this should be okay to do even if the module is not active, although it's settings wouldn't be preloaded, and none of its datapoints would work (although I don't think we need them for this).

What do you think?

**Comment by felixarntz** [comment]
@aaemnnosttv 
1. Makes sense, let's not handle the button adjustments in settings. I.e. the recommendations to Analytics activation should only happen during setup. The settings parts are okay to only cover the rest (matching, informing, blocking the user).
2. For the cases where it's either no AMP or primary AMP, we can follow the logic outlined in the ACs (just need to look at the respective setting). For secondary AMP, let's look at both containers selected:
    * If none of them have GA, fine.
    * If both of them have GA with the same property, fine.
    * Otherwise, block them and say something that they need to set up both containers with the same Analytics property. Exact messaging TBD, but of course shouldn't block this from getting kickstarted.
3. Isn't the datastore loaded anyway since we load the assets for all modules (not just the active ones)? We did this intentionally, since we need to be able to e.g. also render an Analytics CTA component when the module is not active.

**Comment by aaemnnosttv** [comment]
> the recommendations to Analytics activation should only happen during setup

Sounds good.

> For the cases where it's either no AMP or primary AMP, we can follow the logic outlined in the ACs (just need to look at the respective setting)

Yeah, that's what I was thinking.

> For secondary AMP, let's look at both containers selected

Sounds good, that makes sense.

> Isn't the datastore loaded anyway since we load the assets for all modules (not just the active ones)? We did this intentionally, since we need to be able to e.g. also render an Analytics CTA component when the module is not active.

You're right, I thought only the legacy module JS was loaded everywhere for some reason, but the new entrypoints are enqueued by `Screens::enqueue_screen_assets`, so no problem 👍 

---

I'll keep going on the IB then 👍  Would you please update the ACs to reflect these changes?

**Comment by felixarntz** [comment]
@aaemnnosttv Updated the ACs accordingly - IB is ready to be continued!

**Comment by aaemnnosttv** [comment]
@felixarntz I've made some updates here but still WIP.

I think we need to also check `hasTagPermission` with the "detected tag" in GTM, since there is no guarantee that the ID in the container is one that your user has access to as it is entered manually. We should essentially treat this as an existing tag as it would also result in duplicate tracking.

What do you think? I feel like adding another layer of logic here further complicates things but not sure there is a way around it. The problem with this is that if GA is not active, we can't call `GET:tag-permission` to check if the user has access or not since the route won't be registered.

Regarding the settings, are there any other changes here aside from the buttons that can only be done in set up? The IB is getting a bit large on the Setup side, but I think most of the rest of the logic should be the same for Settings, so I want to avoid duplicating anything there.

**Comment by felixarntz** [comment]
@aaemnnosttv IB looks good so far, a few comments:

Maybe we should omit the `hasAnalyticsProperty()` selector and only rely on the `getAnalyticsProperty()` selector. The `error` in the object the latter returns is important to consider; using `hasAnalyticsProperty()` is just a tiny fraction simpler, but the problem here is it doesn't consider the error. The name is slightly off because even if there is an error it could mean that the GTM container has a GA property - just not one that's uniquely identifiable. I'd prefer if we relied purely on `getAnalyticsProperty()`, and that should return an error any time no single property can be identified that applies to both AMP and non-AMP (as applicable). For example, in secondary AMP, if only one of the two containers has a GA property set, this should still return an error because then for the respective other one it wouldn't work.

> GA module is active, and container has a GA tag with non-matching property IDs

This makes sense for GA active; but even for GA inactive, we should also cater for any case where the container GA tag itself has a problem, i.e. anytime `getAnalyticsProperty()` returns an `error` (e.g. the AMP and non-AMP setups don't match).

Regarding `SetupForm`: I'm not sure I follow what exactly you were thinking there from the IB, I don't see any usage of the `getAnalyticsProperty()` selector in there? I think we should rely on that first and foremost, and if there is an `error` block the user before even getting towards GA module-related checks. We can probably just display the `error.message` then (it should have content from the ACs as applicable). Once we know there is no `error` and there is a single GTM GA tag property ID that was detected, we can use that going forward for any other GA module comparisons.

Also thinking just for our own better understanding of the codebase we should use more precise variable names, even if they become long. E.g. a variable like `analyticsTagPropertyID` should probably be called `gtmAnalyticsPropertyID`, which I think is more clear when comparing to `analyticsPropertyID`, which is the actual one from the GA module.

**Comment by aaemnnosttv** [comment]
Thanks @felixarntz I think this is should be finalized now. 😅 

One last thought I have here regarding secondary AMP is if one container has a detected Analytics tag, and the other is set to `container_create` I believe we should treat it the same as if it were a non-matching property ID (that's what I have above). We won't create it with the Analytics tag so we should probably treat it the same as if it was already created with no tag. Would this require alternate messaging or the existing cases would cover it?

**Comment by felixarntz** [comment]
@aaemnnosttv Mostly looks good, one more comment: We may not need `getEffectivePropertyID` in `modules/analytics`? The `propertyID` will be set to the existing tag. If it is outdated and there was another existing tag placed after originally setting up Analytics, I'm not sure we'll want to cover that here (since it's the GTM setup flow)? Because if we do, we open yet another can of worms where, if these two are out of sync we would theoretically need to inform "your Analytics property ID is out of sync with an existing Analytics tag"... I think we should just rely on the existing `getPropertyID`.

> In secondary AMP if both containers have non-matching property IDs, they can be both here as a comma-separated string

I don't think we should do that. Since this is an error case anyway, we're better off returning a false-y `propertyID`, otherwise we'll need to check for both `propertyID && ! error` everywhere which is easy to forget.

Regarding your comment about secondary AMP and one of them being `container_create` while the other one has the flag, I agree this should be flagged as incompatibility between them. Let's use the same wording for now.

**Comment by aaemnnosttv** [comment]
> I'm not sure we'll want to cover that here (since it's the GTM setup flow)? Because if we do, we open yet another can of worms where, if these two are out of sync we would theoretically need to inform "your Analytics property ID is out of sync with an existing Analytics tag"... I think we should just rely on the existing `getPropertyID`.

Good point 😄  - updated.

> I don't think we should do that. Since this is an error case anyway, we're better off returning a false-y `propertyID`, otherwise we'll need to check for both `propertyID && ! error` everywhere which is easy to forget.

Not really everywhere - the `error` is really specific to the secondary AMP case - (although I suppose different errors could be there). I feel like the `propertyID` is like `getExistingTag` and `error` is sort of like `getTagPermission`. If you don't have permission it doesn't mean there isn't a `propertyID`. Some logic will likely only be concerned with the presence of a tag
in GTM while other places will be concerned that there is no error. That all to say that I my thinking here was that `propertyID` should always be truthy if there is at least one tag in GTM, even if they don't match in the case of secondary AMP. It seems odd to just pick one for the value, which is why I had it to include both, but that is a bit odd too. Maybe just `true` ?

> Regarding your comment about secondary AMP and one of them being `container_create` while the other one has the flag, I agree this should be flagged as incompatibility between them. Let's use the same wording for now.

👍

**Comment by felixarntz** [comment]
@aaemnnosttv 
> > I don't think we should do that. Since this is an error case anyway, we're better off returning a false-y `propertyID`, otherwise we'll need to check for both `propertyID && ! error` everywhere which is easy to forget.
> 
> Not really everywhere - the `error` is really specific to the secondary AMP case - (although I suppose different errors could be there). I feel like the `propertyID` is like `getExistingTag` and `error` is sort of like `getTagPermission`. If you don't have permission it doesn't mean there isn't a `propertyID`. Some logic will likely only be concerned with the presence of a tag
> in GTM while other places will be concerned that there is no error. That all to say that I my thinking here was that `propertyID` should always be truthy if there is at least one tag in GTM, even if they don't match in the case of secondary AMP. It seems odd to just pick one for the value, which is why I had it to include both, but that is a bit odd too. Maybe just `true` ?

Hmm I still think this feels overly complicated to use. Why would we care about the property IDs if there are two conflicting ones? I think the function should basically return "the single property ID found", and if there is no single one, it should just be an error because no single one can be determined. Really not a fan of having to check for an error specifically to know whether a function failed, and since we would never need to have both, I think the `propertyID` should just be false-y in this case. (And even if we wanted to know the conflicting IDs, they could rather be included in the error.)

Instead of checking for both `propertyID` and `error`, we can then just check whether `propertyID` is truthy, and only if not we can look at the `error` to display a message.

**Comment by aaemnnosttv** [comment]
> I still think this feels overly complicated to use.

Agreed 😓 

> Why would we care about the property IDs if there are two conflicting ones? I think the function should basically return "the single property ID found", and if there is no single one, it should just be an error because no single one can be determined.

Good point. That sounds good except now `false` or some string would both indicate that a GTM tag exists which is a bit confusing logically.

> Really not a fan of having to check for an error specifically to know whether a function failed, and since we would never need to have both, I think the `propertyID` should just be false-y in this case. (And even if we wanted to know the conflicting IDs, they could rather be included in the error.)
>
> Instead of checking for both `propertyID` and `error`, we can then just check whether `propertyID` is truthy, and only if not we can look at the `error` to display a message.

I'm not really a fan of having to check for an error specifically to know whether a function failed either, but we need to check it in one case or the other. E.g. if `propertyID` was always truthy regardless of matching, then we wouldn't need to check for an error if `! propertyID`.

I don't want to get too hung up on this but I think there is a lot of similarity with existing tags and permissions here, and so I think it makes sense to add a few specific selectors here rather than working with the object from `getAnalyticsProperty` directly - how this object represents the property is a bit of an implementation detail and the fact that it can isn't very clear is a bit of an indication that it probably isn't a good API to use directly as a consumer.

I suggest we add a few specific selectors that are more expressive and self-documenting than the current return from `getAnalyticsProperty`

* `hasAnalyticsProperty`  
returns `true` if there is any Analytics tag in GTM, regardless of matching
* `getAnalyticsPropertyID` (or maybe `getAnalyticsSingularPropertyID`?)  
returns the _singular_ string `propertyID` if matching, `false` if no singular ID can be determined, `null` if none
* `getAnalyticsPropertyError`  
returns the error object if present or `null` if none

I think that covers everything. Maybe that means `getAnalyticsProperty` isn't necessary anymore or would be only internal but I think this would make for code that's easier to read and reason about with less magic values to account for as a consumer.

I also reviewed your previous comment where you preferred to not use `hasAnalyticsProperty` but after considering the alternative it seems like it would make more sense to use specific selectors for this.

I haven't updated the IB here yet so let me know what you think first.

**Comment by felixarntz** [comment]
> > I still think this feels overly complicated to use.
> 
> Agreed 😓
> 
> > Why would we care about the property IDs if there are two conflicting ones? I think the function should basically return "the single property ID found", and if there is no single one, it should just be an error because no single one can be determined.
> 
> Good point. That sounds good except now `false` or some string would both indicate that a GTM tag exists which is a bit confusing logically.

I think that's okay. Since the selector is called `getAnalyticsProperty`, I'd expect it to return a single property or `false`. Whether returning a single property means return one out of many, or whether it means only returning one if there _is only_ one entirely is more of semantic detail I'd say. We should document this behavior in the selector description, I think that's sufficient.

> I'm not really a fan of having to check for an error specifically to know whether a function failed either, but we need to check it in one case or the other. E.g. if `propertyID` was always truthy regardless of matching, then we wouldn't need to check for an error if `! propertyID`.

Not sure I follow. I'd think we only need the `error` for display purposes, while logically, only `propertyID` should indicate whether the function was successful or not. That's why I meant, we'd only have to look at `error` if `! propertyID`, to inform the user about what specifically went wrong (because `propertyID` would not give us that information).

> I suggest we add a few specific selectors that are more expressive and self-documenting than the current return from `getAnalyticsProperty`
> 
> * `hasAnalyticsProperty`
>   returns `true` if there is any Analytics tag in GTM, regardless of matching
> * `getAnalyticsPropertyID` (or maybe `getAnalyticsSingularPropertyID`?)
>   returns the _singular_ string `propertyID` if matching, `false` if no singular ID can be determined, `null` if none
> * `getAnalyticsPropertyError`
>   returns the error object if present or `null` if none
> 
> I think that covers everything. Maybe that means `getAnalyticsProperty` isn't necessary anymore or would be only internal but I think this would make for code that's easier to read and reason about with less magic values to account for as a consumer.

Those selectors make sense to me, but maybe we should name them slightly different. Especially I'd want to make sure that `hasAnalyticsProperty` returning `true` doesn't mean that `getAnalyticsPropertyID` also returns a truthy value. Maybe the following selectors:
1. `hasAnalyticsProperties`
2. `hasSingleAnalyticsProperty`
3. `getSingleAnalyticsPropertyID`
4. `getSingleAnalyticsPropertyError`

We can then keep the original `getAnalyticsProperty` as `getSingleAnalyticsProperty`, mark as private and use it in the above selectors 2.-4. We could even have another internal one `getAnalyticsPropertyIDs`, which is used from `getSingleAnalyticsProperty`.

**Comment by aaemnnosttv** [comment]
> > I'm not really a fan of having to check for an error specifically to know whether a function failed either, but we need to check it in one case or the other. E.g. if `propertyID` was always truthy regardless of matching, then we wouldn't need to check for an error if `! propertyID`.

> Not sure I follow. I'd think we only need the `error` for display purposes, while logically, only `propertyID` should indicate whether the function was successful or not. That's why I meant, we'd only have to look at `error` if `! propertyID`, to inform the user about what specifically went wrong (because `propertyID` would not give us that information).

Right, but if they had no properties in GTM then `propertyID` would be `null` which would also pass for `! propertyID`, so  you would have to specifically check for `propertyID === false` in order to know that it failed/errored without checking `error`. So it is possible to know if it passed or failed by `propertyID` alone but it requires a more specific conditional. It's not a big deal though since this is mostly an internal implementation detail now.

> Those selectors make sense to me, but maybe we should name them slightly different. Especially I'd want to make sure that `hasAnalyticsProperty` returning `true` doesn't mean that `getAnalyticsPropertyID` also returns a truthy value.

`hasAnalyticsProperty` would return `true` if the live container(s) had any Analytics GTM tag.

`getAnalyticsPropertyID` would return the property ID from the found GTM tag(s), only if a single property ID could be determined.

So it's possible `hasAnalyticsProperty` would return `true` and `getAnalyticsPropertyID` could return a truthy or falsy value.

> 1. `hasAnalyticsProperties`

This one is potentially confusing. Would it mean they have any property, or specifically that they have multiple? For that reason I think the following selector negates the need for this one (unless there's another reason we'd want it?).

> 2. `hasSingleAnalyticsProperty`

Sounds good to me, although I prefer "singular" over "single", I can't really say why though; I can go either way on it as both are clear enough I think.

> 3. `getSingleAnalyticsPropertyID`
> 4. `getSingleAnalyticsPropertyError`
> We can then keep the original `getAnalyticsProperty` as `getSingleAnalyticsProperty`, mark as private and use it in the above selectors 2.-4. We could even have another internal one `getAnalyticsPropertyIDs`, which is used from `getSingleAnalyticsProperty`.

Sounds good to me. I think there will be a fair amount of private selector composition necessary for this one.

Let's finalize these selectors - either name is okay with me I'm mostly wondering about `hasAnalyticsProperties`. Let me know and I'll update the IB and then I think this should be good to get started on 🚀

**Comment by felixarntz** [comment]
@aaemnnosttv Agreed regarding `hasAnalyticsProperties` being somewhat confusing too. How about this regarding names:
* ~~`getAnalyticsProperty`~~ &rarr; `getSingleAnalyticsProperty`
* ~~`hasAnalyticsProperty`~~ &rarr; `hasSingleAnalyticsProperty`

If those sound good to you, let's move this forward.

**Comment by aaemnnosttv** [comment]
Sounds good, let's do it! 👍

**Comment by felixarntz** [comment]
@aaemnnosttv After our follow-up discussion, let's do this:
* `getAnalyticsPropertyIDs` (returns array of found property IDs)
* `hasMultipleAnalyticsPropertyIDs` (returns `true` if `getAnalyticsPropertyIDs` has multiple _or_ if `getAnalyticsPropertyIDs` has one while one of the GTM container settings are currently set to `container_create`)
* `getSingleAnalyticsPropertyID` (returns the first item from `getAnalyticsPropertyIDs`, but only if `! hasMultipleAnalyticsPropertyIDs`)
* `hasSingleAnalyticsPropertyID` (returns `!! getSingleAnalyticsPropertyID`)

This way there's no need for any `error` in the return value, and we can also use more precise names referring to property _IDs_.

**Comment by aaemnnosttv** [comment]
IB updated and ✅  as discussed 🚀

**Comment by aaemnnosttv** [comment]
@felixarntz just finishing up implementing the selectors now, so the initial PR for the datastore parts should be ready some time early next week.

A few minor changes to mention and consider here:

I replaced the `hasSingleAnalyticsPropertyID` selector with a `hasAnyAnalyticsPropertyID` instead since `!! getSingleAnalyticsPropertyID` isn't really useful as a selector, especially since `getSingleAnalyticsPropertyID` has 3 falsy return values that mean different things 😄  (`false` - no single/match, `null` both have none, and `undefined`)

One other detail here that may or may not be a change to the original plan is that `getAnalyticsPropertyIDs` now includes `null` in the return value. This means it will always have at least one item (`null`) if no property ID is found. This is important though for `hasMultipleAnalyticsPropertyIDs` in the case that one container has an ID and the other does not - this is similar to the logic for if a container is set to `container_create` - it should count as a unique ID (even though it's not a valid property ID) so that `hasMultipleAnalyticsPropertyIDs` would return `true`. This selector could actually be renamed to `hasMatchingAnalyticsPropertyIDs` (which would have an inverse return value) but that's essentially what has multiple does. Perhaps we should rename this one as I think that makes a bit more sense since the matching part is what we actually care about, (which is only possible with multiple anyways). Just like `hasMultiple` though, `null` would still be a valid value for matching - matching wouldn't inherently mean Analytics property IDs are _present_, only that both containers have the same value.

Other than that, there is a bit of refactoring to do as there is currently quite a bit of duplication in the registry configuration in each test which I plan on drying up a bit into a simple helper. Ideally, I'd like to create a factory for the live container version data as well (everything is currently using several different fixtures) but we'll see about that since the object is kind of a collection of multiple related entities.

**Comment by felixarntz** [comment]
@aaemnnosttv Ready for the components PR!

**Comment by aaemnnosttv** [comment]
@felixarntz @marrrmarrr What do you think about displaying the detected Analytics property ID next to the container if it has one? This would particularly help in the case that's shown here where the IDs don't match but the user has no indication as to which one to change. In this case, one of them doesn't have an Analytics tag set so it would be obvious.

![image](https://user-images.githubusercontent.com/1621608/91849109-32b79c00-ec64-11ea-9140-48411a52e5e5.png)

alternatively something like this 
![image](https://user-images.githubusercontent.com/1621608/91849394-abb6f380-ec64-11ea-8039-418d75b49f92.png)

This is a bit of a prototype so we could make it look a bit more polished, but curious if you feel this would be worth adding (maybe as part of a follow-up issue?). It could be useful to show for non-secondary AMP as well, not just for this particular error state.

**Comment by marrrmarrr** [comment]
@aaemnnosttv this seems like it would be helpful for a lot of users. In terms of positioning, would happen if the user clicks the arrow to open the drop-down menu (right now the ID is immediately underneath) -- would it move farther down?

**Comment by aaemnnosttv** [comment]
@marrrmarrr the contents of the dropdown are layered on top when opened so it would cover the detected Analytics property ID. This would only show for the selected container as we need to query it individually (it would be too much to query them all and show all the Analytics ID in the dropdown for each container option).

As it is, this is a bit of a rough proof-of-concept that I thought of while working on this, but I think the design needs some refinement though. I've done a bit of looking through the material design docs to see if there's anything in there that would be a good fit for this but haven't found anything yet (maybe a tooltip?).

The other challenge here is that there should also be a loading state to indicate that additional container information is being loaded which is necessary before the user can proceed. This isn't actually specific to the display of the Analytics property ID, but something we have to do anyways to know if they can proceed or not. As it is, the submit button is just grayed out for a few seconds after the user selects a container. I thought I had posted a screen cap of this above but looks like that may have been on another issue and I can't find it now. I can take a new one if you'd like to see what I mean.

**Comment by cole10up** [comment]
**Tested**

Installed latest SK, updated branch to PR #2140

Confirmed all green items function as properly.

Validated AMP and Google message with missmatch properties:
![image](https://user-images.githubusercontent.com/57115907/95501432-25758780-095d-11eb-9e33-16de2df0269a.png)

Validated Tag Manager core functionality works as expected and nothing is broken.

Passed QA ✅
